Use ko for local development and releases

ImJasonH opened the pull request:
As discussed in today's community meeting, I wanted to propose adopting `ko` for local development and some release automation.

Motivations include:
* shorter local development feedback loop
* no requirement on a container runtime, meaning images can be built in a container
* release.yaml generation for nightly/regular releases

Notes I shared in the call, for posterity: https://hackmd.io/9RNfkSRDTVuLjUJ583XsOw?view

I tried to find any references I could to a docker- or podman-based development workflow, but I can't guarantee I found them all. Please take a close look and let me know if I missed anything.

Also happy to use this as a forum for any questions/feedback, and obviously if we decide that ko isn't a good fit I can also drop this PR.

sbose78 commented:
/ok-to-test

ImJasonH commented:
[The integration test failure](https://github.com/shipwright-io/build/pull/562/checks?check_run_id=1807720071#step:9:1093) doesn't seem related to this change, but I could be wrong. Does this seem familiar to anyone?

ImJasonH commented:
> [The integration test failure](https://github.com/shipwright-io/build/pull/562/checks?check_run_id=1807720071#step:9:1093) doesn't seem related to this change, but I could be wrong. Does this seem familiar to anyone?

I see this error in the logs a lot, which might be the cause of the e2e test failures: `the object has been modified; please apply your changes to the latest version and try again` -- I don't know how deploying the controller using `ko` would cause this, or if this is a red herring. I'll take a closer look.

(The Travis failure is due to `ko` not being installed in a prior step, I can fix that too)

qu1queee commented:
afaik there is no need for `ko` in integration-tests, we do not use any image in that test, we only instantiate an operator in the background, see [call](https://github.com/shipwright-io/build/blob/6d8f56ee10ab70dd12e4ed7b87784294ea2cbef7/test/integration/utils/operator.go#L17-L40). I will remove ko installation from there. Sorry, I haven´t been able to review this pr.

SaschaSchwarze0 commented on Makefile:
```diff
@@ -93,11 +91,7 @@ build-plain:
-.PHONY: push-image 
-push-image:
-	$(CONTAINER_RUNTIME) push $(IMAGE_HOST)/$(IMAGE):$(TAG)
+	ko publish -P ./cmd/manager
```
The PR [Add trimpath Go build flag #565](https://github.com/shipwright-io/build/pull/565) has added some go flags. Another one is in review: [Tweak build operator build settings #569](https://github.com/shipwright-io/build/pull/569/files). I think there is an environment variable that can be used, [`GOFLAGS`](https://github.com/google/ko#how-can-i-set-ldflags). I guess changing this line to

`GOFLAGS=$(GO_FLAGS) ko publish -P ./cmd/manager`

might be needed.

ImJasonH replied:
Good point. `ko` always adds trimpath if the version of Go supports it: https://github.com/google/ko/blob/v0.8.0/pkg/build/gobuild.go#L317

We can add any other ldflags you'd like, as you said, using GOFLAGS.

SaschaSchwarze0 replied:
Can we use the `GO_FLAGS` that are already in the file ?

```suggestion
	GOFLAGS=$(GO_FLAGS) ko publish -P ./cmd/manager
```

SaschaSchwarze0 commented on hack/release.sh:
```diff
@@ -7,12 +7,12 @@
-make push-image
+# Using defaults, this pushes to:
+# quay.io/shipwright/shipwright-operator/github.com/shipwright-io/build/cmd/manager:latest
+KO_DOCKER_REPO="$IMAGE_HOST/$IMAGE" ko resolve -t "$TAG" -P -R -f deploy/ > release.yaml
```
From the documentation of [`ko resolve`](https://github.com/google/ko#ko-resolve), I am not sure if this step is triggering a build under the cover or only creates the yaml ? In any case, I guess this will need to go into the Makefile to use the GO_FLAGS defined there.

ImJasonH replied:
`ko resolve` scans for `ko://` references, builds those paths, and replaces the placeholders with built image references by digest in the concatenated output YAML, which it sends to stdout (here, to `release.yaml`). So yes, if we want to pass through GOFLAGS we'd need to do that here too.

SaschaSchwarze0 replied:
Yeah, I think we want the `GO_FLAGS` from the makefile to be applied here as well.

SaschaSchwarze0 reviewed (changes requested):
Hi @ImJasonH, thank you for engaging on this topic. It looks like a nice simplification. But, I would like to ask for a staged approach here to not break our downstream builds. There, the convention at the moment is that there is a Dockerfile that runs the build. Changing this to `ko` will need some time. I therefore ask you to keep the Dockerfile in the repo, maybe add a header stating that it is deprecated and will be removed. I think its build will still work as you retain the necessary make targets.

ImJasonH commented:
> Hi @ImJasonH, thank you for engaging on this topic. It looks like a nice simplification. But, I would like to ask for a staged approach here to not break our downstream builds. There, the convention at the moment is that there is a Dockerfile that runs the build. Changing this to `ko` will need some time. I therefore ask you to keep the Dockerfile in the repo, maybe add a header stating that it is deprecated and will be removed. I think its build will still work as you retain the necessary make targets.

Sure, I'd be happy to stage this out however it makes sense. Keeping the Dockerfile as deprecated for now makes sense. Please let us know when you no longer need it so we can clean it up and avoid confusion.

Are the changes to deploy/operator.yaml acceptable to you? If your release workflow is building and replacing the `image:` you should be mostly okay. Removing `command:` means you'll need to make sure the entrypoint is set correctly in the image you build.

SaschaSchwarze0 commented:
> > Hi @ImJasonH, thank you for engaging on this topic. It looks like a nice simplification. But, I would like to ask for a staged approach here to not break our downstream builds. There, the convention at the moment is that there is a Dockerfile that runs the build. Changing this to `ko` will need some time. I therefore ask you to keep the Dockerfile in the repo, maybe add a header stating that it is deprecated and will be removed. I think its build will still work as you retain the necessary make targets.
> 
> Sure, I'd be happy to stage this out however it makes sense. Keeping the Dockerfile as deprecated for now makes sense. Please let us know when you no longer need it so we can clean it up and avoid confusion.
> 
> Are the changes to deploy/operator.yaml acceptable to you? If your release workflow is building and replacing the `image:` you should be mostly okay. Removing `command:` means you'll need to make sure the entrypoint is set correctly in the image you build.

Sounds good, we can remove the Dockerfile once we adjusted our builds. The changes in the operator.yaml are fine. We are not using it.

ImJasonH commented:
Another new flake in integration tests:

```
------------------------------
• Failure [3.282 seconds]
Integration tests Build and BuildRuns
/home/runner/work/build/build/test/integration/build_to_buildruns_test.go:29
  when a build is annotated for deleting the buildrun
  /home/runner/work/build/build/test/integration/build_to_buildruns_test.go:326
    does delete the buildrun after several modifications of the annotation [It]
    /home/runner/work/build/build/test/integration/build_to_buildruns_test.go:426

    Expected
        <[]string | len:0, cap:0>: []
    to contain element matching
        <string>: build-test-build-240

    /home/runner/work/build/build/test/integration/build_to_buildruns_test.go:462
```


And this unrelated flake on Travis 🙄:

```
./hack/install-registry.sh
Unable to find image 'registry:2' locally
2: Pulling from library/registry
anyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit.
```

Retrying 🙏

qu1queee commented:
@ImJasonH im taking notes on the flakiness :)

ImJasonH commented:
Added `GOFLAGS`. Since `ko` always uses `-trimpath`, I've removed it from `GO_FLAGS` in the Makefile and passed it explicitly in `make build` to maintain previous behavior.

HeavyWombat commented:
> Added `GOFLAGS`. Since `ko` always uses `-trimpath`, I've removed it from `GO_FLAGS` in the Makefile and passed it explicitly in `make build` to maintain previous behavior.

I played around with the `ko` branch today and also came across the question regarding the Go build flags. Out curiosity, I was wondering whether `ko` could do it similar to [`goreleaser` and store the build flag details](https://goreleaser.com/customization/build/) in the `.ko.yaml`.

ImJasonH commented:
> I played around with the `ko` branch today and also came across the question regarding the Go build flags. Out curiosity, I was wondering whether `ko` could do it similar to [`goreleaser` and store the build flag details](https://goreleaser.com/customization/build/) in the `.ko.yaml`.

This isn't something `ko` supports today, but it could. Incidentally, `ko` releases itself using goreleaser, and [sets ldflags in its config](https://github.com/google/ko/blob/c2b862d468505dea36ed86e724ca3d190c0d462d/.goreleaser.yml#L15-L16).

I don't personally mind using the default flags for local development, and configuring special flags for released artifacts as necessary. If there are options you want to add it should be easy to do so though.

SaschaSchwarze0 reviewed (approved):
Nice work @ImJasonH.

/lgtm
/approve
